server/src/main/java/org/elasticsearch/index/mapper/PassThroughObjectMapper.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.index.mapper;
+
+import org.elasticsearch.common.Explicit;
+import org.elasticsearch.common.logging.DeprecationLogger;
+import org.elasticsearch.index.IndexVersion;
+import org.elasticsearch.index.mapper.MapperService.MergeReason;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.Locale;
+import java.util.Map;
+
+import static org.elasticsearch.common.xcontent.support.XContentMapValues.nodeBooleanValue;
+
+/**
+ * Mapper for pass-through objects.
+ *
+ * Pass-through objects allow creating fields inside them that can also be referenced directly in search queries.
+ * They also include parameters that affect how nested fields get configured. For instance, if parameter "time_series_dimension"
+ * is set, all keyword subfields are marked as dimensions and included in routing and tsid calculations.
+ */
+public class PassThroughObjectMapper extends ObjectMapper {
+    private static final DeprecationLogger DEPRECATION_LOGGER = DeprecationLogger.getLogger(PassThroughObjectMapper.class);
+
+    public static final String CONTENT_TYPE = "passthrough";
+
+    public static class Builder extends ObjectMapper.Builder {
+
+        // Controls whether subfields are configured as time-series dimensions.
+        protected Explicit<Boolean> containsDimensions = Explicit.IMPLICIT_FALSE;
+
+        public Builder(String name) {
+            // Subobjects are not currently supported.
+            super(name, Explicit.IMPLICIT_FALSE);
+        }
+
+        @Override
+        public PassThroughObjectMapper.Builder add(Mapper.Builder builder) {
+            if (containsDimensions.value() && builder instanceof KeywordFieldMapper.Builder keywordBuilder) {
+                keywordBuilder.dimension(true);
+            }
+            super.add(builder);
+            return this;
+        }
+
+        public PassThroughObjectMapper.Builder setContainsDimensions() {
+            containsDimensions = Explicit.EXPLICIT_TRUE;
+            return this;
+        }
+
+        @Override
+        public PassThroughObjectMapper build(MapperBuilderContext context) {
+            return new PassThroughObjectMapper(name, enabled, dynamic, buildMappers(context.createChildContext(name)), containsDimensions);
```

Comment:
I wonder if we can just iterate over `mappersBuilder` and set dimension to `true` for all builders. We don't need to do that recursively as we force `subobjects` to be `false`. That way, we don't need changes to `MapperBuilderContext` and the `Mapper.Builder#build` method of other mappers.

Reply:
I think this works for fields that are created along with the passthrough object but won't cover the dynamically added dimensions. Note also that this happens in the add() function above.

Reply:
I think you're right that this doesn't cover mapping merges, including dynamic mapping updates.

One option would be to set the `dimension` property of the mappers in the merge result to `true` in `PassThroughObjectMapper#merge`.

Something along the lines of:

```java
mergeResult.mappers()
    .values()
    .stream()
    .map(FieldMapper.class::cast)
    .map(FieldMapper::getMergeBuilder)
    .map(b -> b.dimension(true)) // needs some switch/case to check of instances of field mappers that have a dimension property
    .map(b -> b.build(parentBuilderContext.createChildContext(simpleName())))
    .collect(Collectors.toMap(FieldMapper::simpleName, Function.identity()));
```